torch/csrc/profiler/python/init.cpp

```diff
@@ -326,5 +330,165 @@ void initPythonBindings(PyObject* module) {
   installCapturedTracebackPython();
 }
 
+/* [NOTE: RecordFunctionFast]
+ * These are an alternate way to call record_function from python.
+ * The typical context manager is slow (~15us on benchmarks in Aug 2023), which
+ * is usually fine for module-level annotations in python, but slow for per-op
+ * annotations. Part of the reason it is slow is because the calls go through
+ * the dispatcher, in order to make the record_function calls work with
+ * torchscript.
+ *
+ * This implementation is less safe; it doesn't work with torchscript.
+ *
+ * It implements a context manager in C++, which reduces the overhead compared
+ * to using a python context manager that calls into C++.
```

Comment:
I'm not sure to understand what that means? You mean that having the context manager holding to the guard in python would be slower? How much?

Reply:
[https://github.com/pytorch/pytorch/compare/main...davidberard98:pytorch:record-function-fast-experiments?expand=1](perf experiments) - I was seeing 3.59 us for a python context manager that calls into c++ for `enter` and `exit`; vs 3.12 us for the pure-C++ context manager. (results were pretty noisy, so add +/- .1us).

My suspicion was that the python context manager is slower because it needs to perform additional function calls?